### PR TITLE
Automatically update threshold

### PR DESCRIPTION
**Pull request by klmitch**
This merge request substantially reorganizes the code, utilizing cobra and viper, to allow thresholds to be specified by a configuration file located in the repository.  It also provides logic that allows overcover to automatically determine when the threshold needs to be increased and to update the configuration file, or emit an error if its using a read-only mode (to enable this to be checked in CI pipelines).

### CODE REVIEW

**Comment by klmitch** [comment]
@bleggett, would you like to review? :)

File: README.rst [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+When using a configuration file with Overcover, the Overcover tool can
+automatically update the threshold in the file, to help ensure that
+coverage can never decrease.  This is done using the *headroom*
+options, ``--min-headroom`` and ``--max-headroom``
+(``OVERCOVER_MIN_HEADROOM`` and ``OVERCOVER_MAX_HEADROOM`` for
+environment variables, and ``min_headroom`` and ``max_headroom`` in
+the configuration files).  If the total coverage exceeds the threshold
+plus the maximum headroom, then the threshold will be updated to be
+the total coverage minus the minimum headroom.  An example of a
+configuration file set up to use this automatic update feature would
```
Comment (bleggett) [inline README.rst:80]: This needs a simple "For example, if existing coverage was X and evaluated coverage was Y and min/max headroom was set to P and Q, then the new coverage threshold would be Z" section with example numbers.

File: README.rst [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    ---
+    coverprofile: coverage.out
+    threshold: 75
+    min_headroom: 1
+    max_headroom: 2
+
+Preventing Automatic Update
+---------------------------
+
+Sometimes it is convenient to prevent an automatic update of the
```
Comment (bleggett) [inline README.rst:115]: What's the unconfigured default? 

Touching the file on disk and updating the threshold?

Have a slight pref for making the unconfigured default "touch nothing".

Comment (klmitch) [reply README.rst:115]: Unconfigured default is "touch nothing", actually.  I'll document that :)

File: README.rst (hunk 1 above)
Comment (klmitch) [reply README.rst:80]: Hmm, thought I had some sample numbers in there.  I'll make sure to add an example to describe that.

**Review by bleggett — APPROVED** [review]

**Comment by bleggett** [comment]
LGTM - nice work - will update the repo on coverfail to point this way.

**Pull request merged**